kubernetes/cache.go

```diff
@@ -110,26 +149,135 @@ func initControllers(clientset kube.Interface, refreshDuration time.Duration) ma
+
+	// RBAC API
+	controllers[serviceroleType] = createIstioIndexInformer(c.istioClient.istioRbacApi, serviceroles, c.refreshDuration, namespace)
+	controllers[servicerolebindingType] = createIstioIndexInformer(c.istioClient.istioRbacApi, servicerolebindings, c.refreshDuration, namespace)
+	// controllers[clusterrbacconfigType] = createIstioIndexInformer(c.istioClient.istioRbacApi, clusterrbacconfigs, c.refreshDuration, namespace)
+
+	// Enable
+	c.istioControllers[namespace] = controllers
+}
+
+func createIstioIndexInformer(getter cache.Getter, resourceType string, refreshDuration time.Duration, namespace string) cache.SharedIndexInformer {
+	return cache.NewSharedIndexInformer(cache.NewListWatchFromClient(getter, resourceType, namespace, fields.Everything()),
+		&GenericIstioObject{},
+		refreshDuration,
+		cache.Indexers{},
+	)
+}
+
+func (c *controllerImpl) ProjectListener() cache.ResourceEventHandlerFuncs {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			project := obj.(*osproject_v1.Project)
+			namespace := project.GetObjectMeta().GetName()
+			c.controllerInitializer(namespace)
+		},
+		DeleteFunc: func(obj interface{}) {
+			project := obj.(*osproject_v1.Project)
+			namespace := project.GetObjectMeta().GetName()
+			c.stopNamespace(namespace)
+		},
+	}
+}
+
+func (c *controllerImpl) NamespaceListener() cache.ResourceEventHandlerFuncs {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			namespaceObj := obj.(*core_v1.Namespace)
+			namespace := namespaceObj.GetObjectMeta().GetName()
+			c.controllerInitializer(namespace)
+		},
+		DeleteFunc: func(obj interface{}) {
+			namespaceObj := obj.(*core_v1.Namespace)
+			namespace := namespaceObj.GetObjectMeta().GetName()
+			c.stopNamespace(namespace)
+		},
+	}
+}
+
+func (c *controllerImpl) controllerInitializer(namespace string) {
+	if !c.kubernetesEnabled && !c.istioEnabled {
+		return
+	}
+
+	if _, found := c.stopChans[namespace]; found {
+		// Existing project reappeared.. we might have new settings, so closing the previous ones
+		log.Debugf("Namespace %s reappeared to the informer.. closing existing", namespace)
+		c.stopNamespace(namespace)
+	}
+
+	c.syncLock.Lock()
+	defer c.syncLock.Unlock()
+	// Maybe this should wait for the actual controllers to start before
+	// creating more?
+
+	stopChan := make(chan struct{})
+	c.stopChans[namespace] = stopChan
+	log.Debugf("Starting controllers for namespace: %s\n", namespace)
+	if c.kubernetesEnabled {
+		c.initKubernetesControllersNamespace(namespace)
+		run(c.controllers[namespace], stopChan)
+	}
+	if c.istioEnabled {
+		c.initIstioControllersNamespace(namespace)
+		run(c.istioControllers[namespace], stopChan)
+	}
+
+	// TODO Previous istioEnabled && kubernetesEnabled are allowing isEnabled too soon, before the
+	// caches are synced
 
-	return controllers
+	c.WaitForSync()
```

Comment:
I wonder if we should add a timeout here. The scenario that caches never finished for some reason should not wait forever here, right ?
I mean, if the caches have a problem I guess it can make it crash with some reason sooner than wait in a slow machine.

Just an idea, I know that the previous code was pretty basic and it didn't cover this either.

Reply:
Yeah, probably makes sense. Although so far it has always failed to something else.  This should probably happen on line 310 I think instead of here though.